lib/MatchingAlgorithm.cc

```diff
@@ -388,3 +389,102 @@ std::cout << stmp.str();
+  for (auto &it : proposing_partition) {
+    auto v = it.second;
+    free_list.push(v);
+    int pref_list_size = v->get_preference_list().size();
+    int residual = v->get_upper_quota();
+    bookkeep_data[v] = VertexBookkeeping(0, pref_list_size, 0, residual);
+  }
+
+  // there is at least one vertex in the free list
+  while (not free_list.empty()) {
+    // arbitrary vertex in free list
+    auto u = remove_from_free_list(free_list, bookkeep_data);
+    const auto &u_pref_list = u->get_preference_list();
+    auto &u_data = bookkeep_data[u];
+    // if u^l can have a partner and hasn't exhausted its preference list
+    while (u_data.residual > 0 and not u_data.is_exhausted()) {
+      // Highest ranked vertex to whom u has not yet proposed.
+      VertexPtr v;
+      if (u_pref_list.is_tied(u_data.begin)) {
+        v = u_pref_list.get_ties(u_data.begin)[u_data.tied_index].vertex;
+      } else {
+        v = u_pref_list.at(u_data.begin).vertex;
+      }
+      const auto &v_pref_list = v->get_preference_list();
+
+      if (u_data.level > 0 and M.is_matched_to(v, u, u_data.level - 1)) {
+        // Remove u^(level-1) and v from the matching.
+        M.remove_partner(u, v);
+
+        // Add u^level and v to the matching. Note that residual(u) doesn't
+        // change, as we are replacing one vertex with another.
+        add_matched_partners(M, u, v, u_data, v_pref_list);
+      } else if (M.number_of_partners(v) == v->get_upper_quota()) {
+        // |M[v]| >= upper_quota(v)
+        const auto& v_all_partners = M.get_partners(v);
+        auto v_worst_partner = v_all_partners.get_least_preferred();
+        auto possible_partner =
+            Partner(u, compute_rank(u, v_pref_list), u_data.level);
+
+        if (v_worst_partner < possible_partner) {
+          auto &v_worst_partner_data = bookkeep_data[v_worst_partner.vertex];
+
+          // Increase residual for v's worst partner.
+          v_worst_partner_data.residual += 1;
+
+          // remove M[v_worst_partner] from M[v], and M[v] from
+          // M[v_worst_partner]
+          M.remove_partner(v, v_worst_partner.vertex);
+
+          // add u and v to the matching
+          u_data.residual -= 1;
+          add_matched_partners(M, u, v, u_data, v_pref_list);
+
+          // add v_worst_partner to free_list
+          add_to_free_list(free_list, v_worst_partner_data,
+                           v_worst_partner.vertex);
+        }
+      } else {
+        u_data.residual -= 1;
+        add_matched_partners(M, u, v, u_data, v_pref_list);
+      }
+      if (u_pref_list.is_tied(u_data.begin)) {
+        if(u_data.tied_index < u_pref_list.get_ties(u_data.begin).size() - 1) {
+          u_data.tied_index += 1;
+        } else {
+          u_data.begin += 1;
+          u_data.tied_index = 0;
+        }
+      } else {
+        u_data.begin += 1;
+      }
+    }
+
+    // Activate u^(level+1).
+    if (u_data.residual > 0 and u_data.level < max_level) { // max_level = 0
```

Comment:
What does max_level = 0 mean in this context?

Reply:
It was just a comment for a test case I was trying, I forgot to remove it

Reply:
Ok, sounds good.